### PR TITLE
refactor(datasource): Make "DigestConfig::lookupName" a required field

### LINKED ISSUE #13583: Restrict "datasource.getDigest()" parameters
### Describe the proposed change(s).

Here is how parameters are used by datsource implementations:

|                   | lookupName | registryUrl | currentValue | currentDigest |
| ----------------- | :--------: | :---------: | :----------: | :-----------: |
| aws-machine-image |     ✅     |     ✅      |      ❌      |      ❌       |
| bitbucket-tags    |     ✅     |     ✅      |      ❌      |      ❌       |
| docker            |     ✅     |     ✅      |      ❌      |      ❌       |
| git-refs          |     ✅     |     ❌      |      ❌      |      ❌       |
| git-tags          |     ✅     |     ❌      |      ❌      |      ❌       |
| github-releases   |     ✅     |     ✅      |      ✅      |      ✅       |
| github-tags       |     ✅     |     ✅      |      ❌      |      ❌       |
| gitlab-tags       |     ✅     |     ✅      |      ❌      |      ❌       |
| go                |     ✅     |     ✅      |      ❌      |      ❌       |

Most of them are using `lookupName` and `registryUrl` to form cache key.
We need to make these fields mandatory in `DigestConfig` interface, ensuring their values by helper function before calling `datasource.getDigest()`.
This would allow us to get rid of redundant null guards at the datasource level.
As a result, it should simplify transition to strict null checks.

### PR DESCRIPTION
**Pull request by zharinov**
## Changes:

- Make `lookupName` field mandatory
- Extract config preparation function
- Refactor types
- Refactor tests

## Context:

- Closes: #13583

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/datasource/index.ts [hunk 1]
```diff
@@ -372,24 +373,35 @@ export async function getPkgReleases(
...
+export function supportsDigests(datasource: string | undefined): boolean {
+  return !!datasource && 'getDigest' in getDatasourceFor(datasource);
+}
+
+function getDigestConfig(
+  datasource: DatasourceApi,
+  config: GetDigestInputConfig
+): DigestConfig {
+  const lookupName = config.lookupName ?? config.depName;
+  const [registryUrl = ''] = resolveRegistryUrls(
```
Comment (zharinov) [inline lib/datasource/index.ts:385]: At the moment, I'm not sure about this line

File: lib/datasource/index.ts [hunk 2]
```diff
@@ -372,24 +373,35 @@ export async function getPkgReleases(
...
-  const lookupName = config.lookupName || config.depName;
-  const registryUrls = resolveRegistryUrls(datasource, config.registryUrls);
-  const digestConfig: DigestConfig = {
-    registryUrl: registryUrls[0],
-    currentValue: config.currentValue,
-    currentDigest: config.currentDigest,
-    lookupName,
-  };
-  return datasource.getDigest(digestConfig, value);
+  if (config.datasource) {
```
Comment (viceice) [inline lib/datasource/index.ts:397]: Should be inverted for less code nesting.

**Comment by zharinov** [comment]
It's a way harder to make `registryUrl` required, so let's guarantee only `lookupName` to be provided (well, because it is)

File: lib/datasource/index.ts [hunk 3]
```diff
@@ -372,23 +373,26 @@ export async function getPkgReleases(
   return res;
 }
 
-export function supportsDigests(config: DigestConfig): boolean {
-  return 'getDigest' in getDatasourceFor(config.datasource);
+export function supportsDigests(datasource: string | undefined): boolean {
```
Comment (viceice) [inline lib/datasource/index.ts:376]: Why is this change required?

File: lib/datasource/types.ts [hunk 4]
```diff
@@ -1,13 +1,17 @@
...
+export type GetDigestInputConfig = {
   datasource?: string;
-  depName?: string;
-  lookupName?: string;
   registryUrls?: string[];
-}
+  currentValue?: string;
+  currentDigest?: string;
+} & (
+  | { lookupName: string; depName?: string }
```
Comment (viceice) [inline lib/datasource/types.ts:9]: I think `depName` is always required.

File: lib/datasource/index.ts (hunk 3 above)
Comment (zharinov) [reply lib/datasource/index.ts:376]: Caller of this function passes the type which is now incompatible with `DigestConfig`

File: lib/datasource/types.ts [hunk 5]
```diff
@@ -1,13 +1,16 @@
...
-  depName?: string;
   lookupName?: string;
+  depName: string;
   registryUrls?: string[];
+  currentValue?: string;
+  currentDigest?: string;
 }
 
-export interface DigestConfig extends Config {
+export interface DigestConfig {
```
Comment (rarkins) [inline lib/datasource/types.ts:12]: I've lost track of the difference between this and GetDigestConfig

Comment (zharinov) [reply lib/datasource/types.ts:12]: `GetDigestInputConfig` is raw input, `GetDigest` is sanitized version with resolved values, ready to be consumed from particular datasource. Our problem was that we tried to use same type for sanitized and non-sanitized config, which led to more redundant null-checks inside datasource implementations.

**Review by rarkins — APPROVED** [review]

**Comment by rarkins** [comment]
Please confirm if it's tested on some real repos just to make sure? :)

**Comment by zharinov** [comment]
I confirm 👍

**Review by viceice — APPROVED** [review]
